example/advection/advection.cpp

```diff
@@ -0,0 +1,399 @@
+
+
+  auto start_recv = AddContainerTask([](Container<Real>& rc) {
+    rc.StartReceiving(parthenon::BoundaryCommSubset::all);
+    return TaskStatus::success;
+  }, none, sc1);
+
+  auto advect_flux = AddContainerTask(Advection::CalculateFluxes, none, sc0);
+
+  auto send_flux = AddContainerTask([](Container<Real>& rc) {
+    rc.SendFluxCorrection();
+    return TaskStatus::success;
+  }, advect_flux, sc0);
+
+  auto recv_flux = AddContainerTask([](Container<Real>& rc) {
+      if (!rc.ReceiveFluxCorrection()) return TaskStatus::fail;
+      return TaskStatus::success;
+    }, advect_flux, sc0);
+
+  // compute the divergence of fluxes of conserved variables
+  auto flux_div = AddTwoContainerTask([](Container<Real>& u, Container<Real>& du) {
+    parthenon::Update::FluxDivergence(u, du);
+    return TaskStatus::success;
+  }, recv_flux, sc0, dudt);
+
+  // apply du/dt to all independent fields in the container
+  auto update_container = AddMyTask(UpdateContainer, flux_div);
+
+  // update ghost cells
+  auto send = AddContainerTask([](Container<Real>& rc) {
+    rc.SendBoundaryBuffers();
+    return TaskStatus::success;
+  }, update_container, sc1);
+  auto recv = AddContainerTask([](Container<Real>& rc) {
+    if ( !rc.ReceiveBoundaryBuffers() ) return TaskStatus::fail;
+    return TaskStatus::success;
+  }, send, sc1);
+  auto setC = AddContainerTask([](Container<Real>& rc) {
+    rc.SetBoundaries();
+    rc.ClearBoundary(parthenon::BoundaryCommSubset::all);
+    return TaskStatus::success;
+  }, recv, sc1);
+  auto prolongBound = AddContainerTask([](Container<Real>& rc) {
+    MeshBlock *pmb = rc.pmy_block;
+    BoundaryValues *pbval = pmb->pbval.get();
+    pbval->ProlongateBoundaries(0.0, 0.0);
+    return TaskStatus::success;
+  }, setC, sc1);
+
+  // set physical boundaries
+  auto set_bc = AddContainerTask([](Container<Real>& rc) {
+    parthenon::ApplyBoundaryConditions(rc);
+    return TaskStatus::success;
+  }, prolongBound, sc1);
+
+  // fill in derived fields
+  auto fill_derived = AddContainerTask([](Container<Real>& rc) {
+    parthenon::FillDerivedVariables::FillDerived(rc);
+    return TaskStatus::success;
+  }, set_bc, sc1);
+
+  // estimate next time step
+  if (stage == integrator->_nstages) {
+    auto new_dt = AddContainerTask([](Container<Real>& rc) {
+      MeshBlock *pmb = rc.pmy_block;
+      pmb->SetBlockTimestep(parthenon::Update::EstimateTimestep(rc));
+      return TaskStatus::success;
+    }, fill_derived, sc1);
+
+    // Update refinement
+    if (pmesh->adaptive) {
+      auto tag_refine = tl.AddTask<BlockTask>([](MeshBlock *pmb) {
+        pmb->pmr->CheckRefinementCondition();
+        return TaskStatus::success;
+      }, fill_derived, pmb);
+    }
+    // Purge stages
+    auto purge_stages = tl.AddTask<BlockTask>([](MeshBlock *pmb) {
+      pmb->real_containers.PurgeNonBase();
+      return TaskStatus::success;
+    }, fill_derived, pmb);
+
+  }
+  return tl;
```

Comment:
It occurs to me that a lot of tasks in the task list are going to look identical for a lot of codes, which is why, of course, we provide methods in `parthenon` to perform them. But I wonder if there could be a method in `MultiStageIntegrator` to add all of them at once?

Reply:
problem is dependencies.  how do these commonly performed tasks fit in with the rest of the stuff that needs to happen?

Reply:
I feel like at least some of them depend on nothing, one unknown task, or only each other, which might be consistent across a wide range of methods. So there might be something that can be done to stitch them in piecemale. Whether or not that's worth doing, I don't know. It's probably a question for another time.

Reply:
@Yurlungur is it possible to provide pre-canned tasks in multistage integrator, similar to the methods in parthenon.  This also opens the door to behind-the-scenes optimization for common tasks.  I actually like the specification of the dependencies in the place where they are referenced and wouldn't like to muck that up.

Reply:
Ooh, I like that @nmsriram. Maybe we could provide methods that let you do something like
```C++
auto recv = AddRecvGhostsTask(dependency);
```

Reply:
@nmsriram @Yurlungur Not sure how to get exactly that API, but I have been meaning to make tasks out of a lot of the generic parthenon functionality.  I've pushed a step in this direction, providing tasks for Container mediated MPI ops.  I'm not sure the static member function approach I took is necessarily the best, but I don't hate it.  Thoughts?

Reply:
I like that, @jdolence. I'll take a more careful look at the API as you implemented.